### PR TITLE
Add AMX RTB adapter

### PR DESCRIPTION
**Pull request by nickjacob**
Adding AMX RTB adapter

Server uses standard OpenRTB 2.5. This adapter adds support for overriding the publisher id using `imp.ext.amx.tagId`, although this is optional. The parameter is named to match the bid param of [our prebid.js adapter](https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/amx-rtb.md)

Maintainer email: prebid@amxrtb.com

### CODE REVIEW

**Comment by bretg** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/2437

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:61]: With the other suggested change, this would become:
```
if publisherID, ok := getPublisherID(request.Imp); ok {
	if request.App != nil {
		request.App.Publisher = ensurePublisherWithID(request.App.Publisher, publisherID)
	}
	if request.Site != nil {
		request.Site.Publisher = ensurePublisherWithID(request.Site.Publisher, publisherID)
	}
}
```

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:44]: There is no consequence to this print statement, neither in the response or the in the log. The chance of this happening is also likely 0% due to preprocessing by the "core" logic. I recommend removing the print statement since the publisher id is optional for your adapter.

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:71]: Nitpick: In Go, it's common to remove the line break after the `json.Marshal` call to group it with the error handling code.

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:82]: Please consider adding this query parameter safely using Go's url methods within the `NewAmxBidder` method.

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:105]: Since this error would likely mean the server responded back with a bad response, it's common to use `BadServerResponse` here instead.

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:152]: Does your bid response include the bid type? Although the code you have here is common in the project, there's been issues reported (see https://github.com/prebid/prebid-server/issues/862) that this approach isn't always reliable for multiformat impressions. Is there enough information to detect the type based on the bid response?

File: adapters/amx/usersync.go
Comment (SyntaxNode) [inline adapters/amx/usersync.go:12]: GVL ID is registered to `Monet Engine Inc`. Is that correct for AMX?

File: adapters/amx/usersync_test.go
Comment (SyntaxNode) [inline adapters/amx/usersync_test.go:24]: Please follow the format form the other syncer tests.

File: usersync/usersyncers/syncer.go
Comment (SyntaxNode) [inline usersync/usersyncers/syncer.go:103]: Per Go naming convention, this should be `BidderAMX`, `NewAMXSyncer`, etc...

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:32]: Please move `amxParams` to the `openrtb_ext` package in the file openrtb_ext/imp_amx.go with the name `ExtImpAmx` to match our standard convention.

**Review by SyntaxNode — COMMENTED** [review]
Thank you for your contribution. Very glad to see the docs PR already open. :)

**Comment by nickjacob** [comment]
Sorry for some reason github won't let me resolve the comments or reply 
here—but thanks for the great CR!

I made the changes you noted. Our bids do always have `startdelay` in 
bid.ext when video, so I used that instead to determine the type.

For the GVL—737 is correct thanks

I renamed everything to AMX (from Amx). Let me know if there's anything 
else

Thanks!

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:119]: There's a problem here because if `bid.Ext` is empty the `json.Unmarshal` call will return an error and not produce a bid response. This doesn't fail the exemplary tests because you're missing `expectedBidResponses` from those files.

This will need to be changed to handle a missing `bid.Ext`. Perhaps something like:
```
for _, sb := range bidResp.SeatBid {
		for _, bid := range sb.Bid {

			bidExt, bidExtErr := getBidExt(bid.Ext)
			if bidExtErr != nil {
				errs = append(errs, bidExtErr)
				continue
			}

			bidType := getMediaTypeForBid(bidExt)

			b := &adapters.TypedBid{
				Bid:     &bid,
				BidType: bidType,
			}
			if b.BidType == openrtb_ext.BidTypeVideo {
				b.Bid.AdM = interpolateImpressions(bid, bidExt)
			}
			bidResponse.Bids = append(bidResponse.Bids, b)

		}
	}
```

```
func getBidExt(ext json.RawMessage) (amxBidExt, error) {
	if len(ext) == 0 {
		return amxBidExt{}, nil
	}

	var bidExt amxBidExt
	err := json.Unmarshal(ext, &bidExt)

	return bidExt, err
}
```

File: adapters/amx/amx_test.go
Comment (SyntaxNode) [inline adapters/amx/amx_test.go:74]: Please consider using testify assert statements in your tests. Not strictly required. A bit easier to read though and you don't need to be concerned with edge cases.

```
assert.NoError(t, actualAdapterRequests, "request")
assert.Empty(t, err, "request errors")
...
assert.Equal(t, "publisher_id_example", body.Site.Publisher.ID, "publisher")
```

File: adapters/amx/amx_test.go
Comment (SyntaxNode) [inline adapters/amx/amx_test.go:102]: It's not valid for a BidRequest to contain both a `Site` and `App` section. This will be caught earlier on by the framework. For this test, please only use the `App`.

File: adapters/amx/amx_test.go
Comment (SyntaxNode) [inline adapters/amx/amx_test.go:206]: Please fix the whitespace on these lines for consistency.

File: adapters/amx/amx_test.go
Comment (SyntaxNode) [inline adapters/amx/amx_test.go:254]: Same here. Please use consistent whitespace for the json blobs.

File: adapters/amx/amx_test.go
Comment (SyntaxNode) [inline adapters/amx/amx_test.go:351]: assert.Contains(t, data, "example.com/pixel.png", "interpolated into vast")

File: adapters/amx/amxtest/exemplary/app.simple.json
Comment (SyntaxNode) [inline adapters/amx/amxtest/exemplary/app.simple.json:152]: Please add expectedBidResponses to the json test files.

**Review by SyntaxNode — COMMENTED** [review]
Thank you for the quick update from our feedback. I took a closer look into the tests and found a few items to address.

**Comment by nickjacob** [comment]
Thanks @SyntaxNode  -- I made the requested CR changes; refactored to use assert which did really improve the tests, adding some more test coverage and "supplemental" JSON tests, expectedBidResponse & expectedMakeBidsErrors in JSON tests

File: adapters/amx/amx_test.go
Comment (mansinahar) [inline adapters/amx/amx_test.go:281]: Not exactly sure what exactly this test is testing because I don't see any modifications around user ext in your `MakeRequests` function. This to me looks more like making sure that user exts are not getting modified by `MakeRequests`. Is that correct? 

Also, I'd suggest adding these tests in this file as part of the JSON tests rather than unit tests as that will help remove a lot of setup for your tests which is covered by `RunJSONBidderTest` and will also provide consistency in tests with other adapters. In fact some of these test cases are already covered with the JSON tests.

File: static/bidder-params/amx.json
Comment (mansinahar) [inline static/bidder-params/amx.json:7]: Bidder params should have the params expected by your adapter which in your case should be `tagid`. You will have to update the `adapters/amx/amxtest/params/race/display.json` and `adapters/amx/amxtest/params/race/video.json` file accordingly as well.

**Comment by nickjacob** [comment]
Thanks @mansinahar -- I made those changes & rebased against master

File: adapters/amx/amx_test.go
Comment (mansinahar) [reply adapters/amx/amx_test.go:26]: @nickjacob Could you please address this? looks like this was missed in the update. Thanks!

Comment (nickjacob) [reply adapters/amx/amx_test.go]: sorry about that missed this -- just pushed another commit + with the table-driven test for the publisher ID logic

**Review by mansinahar — COMMENTED** [review]
Please add a `params_test.go` file as well to test the bidder params. For reference you can check this one for 33across: https://github.com/prebid/prebid-server/blob/master/adapters/33across/params_test.go

File: adapters/amx/amx.go
Comment (SyntaxNode) [inline adapters/amx/amx.go:35]: Please handle the error from `url.ParseQuery` in a similar manner as you did with `url.Parse`.

**Comment by nickjacob** [comment]
added params_test.go, improved testing of interpolation function + better error handling in `NewAMXBidder`. Also added default sync endpoint

also rebased off of master just to be safe

**Review by mansinahar — DISMISSED** [review]

**Comment by SyntaxNode** [comment]
@nickjacob Can you please address my comment on the docs PR?

**Review by SyntaxNode — APPROVED** [review]
I'm good to merge this in. The docs PR is open with one remaining comment, but I don't see a reason for that to be a blocker here.

**Comment by nickjacob** [comment]
@SyntaxNode  -- for the docs PR; we don't want to have all those same parameters for PBS that we have in PBJS (e.g. testMode is the same as test: 1, so not necessary). Should I make a second table in the doc that describes the PBS parameters separate from PBJS?

**Comment by SyntaxNode** [comment]
> @SyntaxNode -- for the docs PR; we don't want to have all those same parameters for PBS that we have in PBJS (e.g. testMode is the same as test: 1, so not necessary). Should I make a second table in the doc that describes the PBS parameters separate from PBJS?

Yes, please. The `testMode` makes sense. 

Before I git the merge button... what about `tagId` vs `publisher_id` and the `adUnitId`?

**Comment by nickjacob** [comment]
I ended up changing the parameter here to be "tagId" to match "tagId" in PBJS. adUnitId is probably a good idea you're right. I'll add that now so the PBS version better matches PBJS

**Comment by SyntaxNode** [comment]
> I ended up changing the parameter here to be "tagId" to match "tagId" in PBJS. adUnitId is probably a good idea you're right. I'll add that now so the PBS version better matches PBJS

Ah, right. Sorry. Forgot about the `tagId` change was already made. I'll hold off on merging for the `adUnitId` then. Thank you.

**Review by SyntaxNode — APPROVED** [review]

**Review by mansinahar — APPROVED** [review]

**Pull request merged**
